File: protocol/src/messages.ts
import { ChannelMeta, RoleMeta, ServerMeta, UserPresence, InviteInfo } from './types';

export * from './types';

export interface ClientMessage {
  id: string;
  version: number;
  timestamp: number;
  sender: string;
  type: string;
  payload: unknown;
}

export interface ServerMessage {
  id: string;
  version: number;
  timestamp: number;
  sender: string;
  type: string;
  payload: unknown;
}

export interface EncryptedPayload {
  ciphertext: string;
  nonce: string;
  sender_key: string;
  recipient_key?: string;
  signature: string;
}

export interface ChatMessage {
  type: 'chat.message';
  channel_id: string;
  ciphertext: string;
  nonce: string;
  reply_to?: string;
  thread_id?: string;
}

export interface ChatEdit {
  type: 'chat.edit';
  channel_id: string;
  message_id: string;
  ciphertext: string;
  nonce: string;
}

export interface ChatDelete {
  type: 'chat.delete';
  channel_id: string;
  message_id: string;
}

export interface ChatReaction {
  type: 'chat.reaction';
  channel_id: string;
  message_id: string;
  emoji: string;
  action: 'add' | 'remove';
}

export interface ChannelCreate {
  type: 'channel.create';
  name: string;
  topic?: string;
  channel_type: 'text' | 'voice';
}

export interface ChannelUpdate {
  type: 'channel.update';
  channel_id: string;
  name?: string;
  topic?: string;
}

export interface ChannelDelete {
  type: 'channel.delete';
  channel_id: string;
}

export interface PresenceUpdate {
  type: 'presence.update';
  status: 'online' | 'idle' | 'dnd' | 'offline';
}

export interface SyncRequest {
  type: 'sync.request';
  since?: number;
  channel_id?: string;
}

export type InboundMessage =
  | ChatMessage
  | ChatEdit
  | ChatDelete
  | ChatReaction
  | ChannelCreate
  | ChannelUpdate
  | ChannelDelete
  | PresenceUpdate
  | SyncRequest;

export interface ChatMessageReceived {
  type: 'chat.message';
  id: string;
  channel_id: string;
  sender: string;
  ciphertext: string;
  nonce: string;
  timestamp: number;
  reply_to?: string;
  thread_id?: string;
}

export interface ChatEditReceived {
  type: 'chat.edit';
  channel_id: string;
  message_id: string;
  ciphertext: string;
  nonce: string;
  timestamp: number;
}

export interface ChatDeleteReceived {
  type: 'chat.delete';
  channel_id: string;
  message_id: string;
  timestamp: number;
}

export interface ReactionReceived {
  type: 'chat.reaction';
  channel_id: string;
  message_id: string;
  emoji: string;
  user_key: string;
  action: 'add' | 'remove';
}

export interface ChannelListReceived {
  type: 'channel.list';
  channels: ChannelMeta[];
}

export interface SyncResponseReceived {
  type: 'sync.response';
  channels: ChannelMeta[];
  pinned_messages: PinnedMessage[];
  timestamp: number;
}

export interface ErrorReceived {
  type: 'error';
  code: string;
  message: string;
}

export type OutboundMessage =
  | ChatMessageReceived
  | ChatEditReceived
  | ChatDeleteReceived
  | ReactionReceived
  | ChannelListReceived
  | SyncResponseReceived
  | ErrorReceived;

export interface PinnedMessage {
  id: string;
  channel_id: string;
  encrypted_blob: string;
  pinned_by: string;
  pinned_at: number;
}

export interface FederationMessage {
  action: 'trust_request' | 'trust_confirm' | 'trust_revoke';
  server_pubkey: string;
  signature: string;
}

export interface FederationResponse {
  status: 'pending_approval' | 'established' | 'revoked';
  message?: string;
  server_pubkey?: string;
  signature: string;
}

export interface RelayRequest {
  source_server: string;
  target_server: string;
  encrypted_payload: string;
  signature: string;
  forward: boolean;
}

File: server/src/protocol/messages.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum ClientMessage {
    #[serde(rename = "chat.message")]
    ChatMessage {
        channel_id: String,
        ciphertext: String,
        nonce: String,
        reply_to: Option<String>,
        thread_id: Option<String>,
    },
    
    #[serde(rename = "chat.edit")]
    ChatEdit {
        channel_id: String,
        message_id: String,
        ciphertext: String,
        nonce: String,
    },
    
    #[serde(rename = "chat.delete")]
    ChatDelete {
        channel_id: String,
        message_id: String,
    },
    
    #[serde(rename = "chat.reaction")]
    ChatReaction {
        channel_id: String,
        message_id: String,
        emoji: String,
        action: ReactionAction,
    },
    
    #[serde(rename = "channel.create")]
    ChannelCreate {
        name: String,
        topic: Option<String>,
        channel_type: ChannelType,
        parent_id: Option<String>,
    },
    
    #[serde(rename = "channel.update")]
    ChannelUpdate {
        channel_id: String,
        name: Option<String>,
        topic: Option<String>,
    },
    
    #[serde(rename = "channel.delete")]
    ChannelDelete {
        channel_id: String,
    },
    
    #[serde(rename = "federation.relay")]
    FederationRelay {
        target_server: String,
        encrypted_payload: Vec<u8>,
    },
    
    #[serde(rename = "sync.request")]
    SyncRequest {
        since: Option<u64>,
        channel_id: Option<String>,
    },
    
    #[serde(rename = "presence.update")]
    PresenceUpdate {
        status: PresenceStatus,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
pub enum ServerMessage {
    #[serde(rename = "chat.message")]
    ChatMessage {
        id: String,
        channel_id: String,
        sender: String,
        ciphertext: String,
        nonce: String,
        timestamp: u64,
        reply_to: Option<String>,
        thread_id: Option<String>,
    },
    
    #[serde(rename = "chat.edit")]
    ChatEdit {
        id: String,
        channel_id: String,
        message_id: String,
        ciphertext: String,
        nonce: String,
        timestamp: u64,
    },
    
    #[serde(rename = "chat.delete")]
    ChatDelete {
        channel_id: String,
        message_id: String,
        timestamp: u64,
    },
    
    #[serde(rename = "channel.list")]
    ChannelList {
        channels: Vec<super::types::ChannelMeta>,
    },
    
    #[serde(rename = "sync.response")]
    SyncResponse {
        channels: Vec<super::types::ChannelMeta>,
        pinned_messages: Vec<PinnedMessage>,
        timestamp: u64,
    },
    
    #[serde(rename = "error")]
    Error {
        code: String,
        message: String,
    },
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ChannelType {
    Text,
    Voice,
    Category,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ReactionAction {
    Add,
    Remove,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum PresenceStatus {
    Online,
    Idle,
    Dnd,
    Offline,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PinnedMessage {
    pub id: String,
    pub channel_id: String,
    pub encrypted_blob: String,
    pub pinned_by: String,
    pub pinned_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FederatedMessage {
    pub action: String,
    pub server_pubkey: Option<String>,
    pub signature: String,
    pub payload: Option<Vec<u8>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FederationMessage {
    pub action: String,
    pub server_pubkey: Option<String>,
    pub signature: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FederationResponse {
    pub status: String,
    pub message: Option<String>,
    pub server_pubkey: Option<String>,
    pub signature: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RelayRequest {
    pub source_server: String,
    pub target_server: String,
    pub encrypted_payload: Vec<u8>,
    pub signature: String,
    pub forward: bool,
}
